File: services/blockchain/src/workers/confirmationPoller.ts
import { Pool } from 'pg'

const STUCK_TX_MS = 24 * 60 * 60 * 1000  // 24 hours

const MEMPOOL_TX_URL: Record<'mainnet' | 'testnet', string> = {
  mainnet: 'https://mempool.space/api/tx',
  testnet: 'https://mempool.space/testnet/api/tx',
}

const NOT_FOUND = Symbol('NOT_FOUND')
const TRANSIENT_ERROR = Symbol('TRANSIENT_ERROR')

interface TxConfirmStatus {
  confirmed: boolean
  block_height?: number
}

async function fetchTxStatus(
  txid: string,
  network: 'mainnet' | 'testnet',
): Promise<TxConfirmStatus | typeof NOT_FOUND | typeof TRANSIENT_ERROR> {
  try {
    const res = await fetch(`${MEMPOOL_TX_URL[network]}/${txid}`, {
      signal: AbortSignal.timeout(10000),
    })
    if (res.status === 404) return NOT_FOUND
    if (!res.ok) return TRANSIENT_ERROR
    const data = await res.json() as { status: TxConfirmStatus }
    return data.status
  } catch {
    return TRANSIENT_ERROR
  }
}

export async function pollOnce(pool: Pool, network: 'mainnet' | 'testnet'): Promise<void> {
  const utxoResult = await pool.query<{ id: string; txid: string; created_at: string }>(
    `SELECT id, txid, created_at FROM utxos WHERE status = 'UNCONFIRMED'`,
  )
  const jobResult = await pool.query<{ id: string; txid: string; source_type: string; source_id: string }>(
    `SELECT id, bitcoin_txid AS txid, source_type, source_id
     FROM publish_jobs WHERE status = 'BITCOIN_ANCHORED' AND bitcoin_txid IS NOT NULL`,
  )

  const utxosByTxid = new Map<string, typeof utxoResult.rows>()
  for (const row of utxoResult.rows) {
    const existing = utxosByTxid.get(row.txid) ?? []
    utxosByTxid.set(row.txid, [...existing, row])
  }

  const jobsByTxid = new Map<string, typeof jobResult.rows>()
  for (const row of jobResult.rows) {
    const existing = jobsByTxid.get(row.txid) ?? []
    jobsByTxid.set(row.txid, [...existing, row])
  }

  const allTxids = new Set([...utxosByTxid.keys(), ...jobsByTxid.keys()])

  for (const txid of allTxids) {
    const status = await fetchTxStatus(txid, network)

    if (status === TRANSIENT_ERROR) {
      console.warn('[poller] transient error fetching txid, will retry next tick:', txid)
      continue
    }

    if (status === NOT_FOUND) {
      console.error('[poller] txid not found on network (404):', txid)
      for (const job of jobsByTxid.get(txid) ?? []) {
        await pool.query(
          `UPDATE publish_jobs
           SET status = 'FAILED', error_message = $2, retry_count = retry_count + 1,
               next_retry_at = NOW(), worker_id = NULL, locked_at = NULL, updated_at = NOW()
           WHERE id = $1`,
          [job.id, `txid not found on network: ${txid}`],
        )
        await pool.query(
          `INSERT INTO publish_failures (job_id, step, error_message) VALUES ($1, 'BITCOIN_BROADCAST', $2)`,
          [job.id, `txid not found: ${txid}`],
        )
      }
      continue
    }

    // Check for stuck transactions (>24h UNCONFIRMED, not yet confirmed)
    if (!status.confirmed) {
      for (const utxo of utxosByTxid.get(txid) ?? []) {
        const ageMs = Date.now() - new Date(utxo.created_at).getTime()
        if (ageMs > STUCK_TX_MS) {
          console.warn('[poller] stuck transaction >24h unconfirmed:', txid)
        }
      }
      continue
    }

    const blockHeight = status.block_height!

    for (const utxo of utxosByTxid.get(txid) ?? []) {
      await pool.query(
        `UPDATE utxos SET status = 'CONFIRMED', updated_at = NOW() WHERE id = $1 AND status = 'UNCONFIRMED'`,
        [utxo.id],
      )
    }

    for (const job of jobsByTxid.get(txid) ?? []) {
      await pool.query(
        `UPDATE publish_jobs SET status = 'COMPLETE', updated_at = NOW()
         WHERE id = $1 AND status = 'BITCOIN_ANCHORED'`,
        [job.id],
      )
      const sourceTable = job.source_type === 'SAFETY_EVENT' ? 'safety_events' : 'community_reports'
      await pool.query(
        `UPDATE ${sourceTable} SET bitcoin_block = $2 WHERE id = $1`,
        [job.source_id, blockHeight],
      )
    }
  }
}

export function startConfirmationPoller(
  pool: Pool,
  network: 'mainnet' | 'testnet',
): { stop: () => void } {
  const timer = setInterval(
    () => pollOnce(pool, network).catch(err => console.error('[poller] tick error:', err)),
    60_000,
  )
  pollOnce(pool, network).catch(err => console.error('[poller] initial tick error:', err))
  return { stop: () => clearInterval(timer) }
}

File: services/blockchain/src/workers/confirmation_poller.rs
// services/blockchain/src/workers/confirmation_poller.rs
use anyhow::{anyhow, Result};
use sqlx::PgPool;
use std::sync::Arc;
use tokio::time::{interval, Duration};
use uuid::Uuid;

use crate::config::Config;
use crate::db::jobs;

#[derive(serde::Deserialize)]
struct MempoolTxStatus {
    confirmed: bool,
    block_height: Option<i32>,
}

#[derive(serde::Deserialize)]
struct MempoolTxResponse {
    status: MempoolTxStatus,
}

#[derive(Debug, sqlx::FromRow)]
struct AnchoredJob {
    id: Uuid,
    source_type: String,
    source_id: Uuid,
    bitcoin_txid: String,
    retry_count: i32,
}

pub async fn run(pool: Arc<PgPool>, config: Arc<Config>) {
    let mut ticker = interval(Duration::from_secs(60));
    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(10))
        .build()
        .expect("failed to build reqwest client for confirmation poller");
    loop {
        ticker.tick().await;
        if let Err(e) = poll_confirmations(&pool, &config, &client).await {
            tracing::error!("confirmation poller error: {}", e);
        }
    }
}

async fn poll_confirmations(pool: &PgPool, config: &Config, client: &reqwest::Client) -> Result<()> {
    let jobs = sqlx::query_as::<_, AnchoredJob>(
        "SELECT id, source_type, source_id, bitcoin_txid, retry_count FROM publish_jobs WHERE status = 'BITCOIN_ANCHORED' LIMIT 50",
    )
    .fetch_all(pool)
    .await?;

    for job in jobs {
        if let Err(e) = check_and_confirm(pool, client, config, &job).await {
            tracing::warn!("confirmation check failed for job {}: {}", job.id, e);
        }
    }
    Ok(())
}

async fn check_and_confirm(
    pool: &PgPool,
    client: &reqwest::Client,
    config: &Config,
    job: &AnchoredJob,
) -> Result<()> {
    let url = config.bitcoin_network.mempool_tx_url(&job.bitcoin_txid);
    let response = client.get(&url).send().await?;

    let http_status = response.status();

    // 404 means the txid is unknown to the network — definitive failure, not transient.
    if http_status == reqwest::StatusCode::NOT_FOUND {
        let error_message = format!("txid not found on network (404): {}", job.bitcoin_txid);
        tracing::error!("job {} txid {} returned 404 — marking FAILED", job.id, job.bitcoin_txid);
        jobs::mark_failed(pool, job.id, &error_message, job.retry_count).await?;
        return Ok(());
    }

    // Any other non-2xx (5xx, other 4xx) is a transient API error — skip this cycle.
    if !http_status.is_success() {
        return Ok(());
    }

    let resp: MempoolTxResponse = response.json().await?;

    if !resp.status.confirmed {
        return Ok(());
    }

    // mempool.space guarantees block_height is present when confirmed=true; treat absence as unexpected.
    let block_height = match resp.status.block_height {
        Some(h) => h,
        None => {
            tracing::warn!("txid {} confirmed but block_height missing in API response, skipping", job.bitcoin_txid);
            return Ok(());
        }
    };

    // table is bounded to two compile-time-known values — no SQL injection risk.
    let table = match job.source_type.as_str() {
        "SAFETY_EVENT" => "safety_events",
        "COMMUNITY_REPORT" => "community_reports",
        other => {
            tracing::warn!("unknown source_type '{}' for job {}, skipping source table update", other, job.id);
            return Err(anyhow!("unknown source_type: {}", other));
        }
    };

    // Wrap both writes in a transaction so the job is never COMPLETE without the source table updated.
    let mut tx = pool.begin().await?;
    sqlx::query("UPDATE publish_jobs SET status = 'COMPLETE', updated_at = NOW() WHERE id = $1")
        .bind(job.id)
        .execute(&mut *tx)
        .await?;
    sqlx::query(&format!(
        "UPDATE {} SET bitcoin_block = $2 WHERE id = $1 AND bitcoin_block IS NULL",
        table
    ))
    .bind(job.source_id)
    .bind(block_height)
    .execute(&mut *tx)
    .await?;
    tx.commit().await?;

    tracing::info!("job {} confirmed in block {} (txid: {})", job.id, block_height, job.bitcoin_txid);
    Ok(())
}
